gplugins/lumerical/eme.py

```diff
@@ -0,0 +1,1063 @@
+        }
+
+        return coupling_coefficients
+
+    def plot_mode_coupling(
+        self, input_mode: int, max_coupled_mode: int = None, group: int = 1
+    ) -> None:
+        """
+        Plot mode coupling coefficients from the input_mode to the max_coupled_mode.
+        Mode numbers start from 1 (fundamental mode).
+
+        Parameters:
+            input_mode: Mode of interest. Energy couples away from this mode to other modes.
+            max_coupled_mode: Maximum mode number to consider; all modes below this mode will be considered as well.
+            group: Group of cells to consider. First group is 0.
+        """
+        coupling_coefficients = self.get_mode_coupling(
+            input_mode, max_coupled_mode, group
+        )
+
+        fig, axs = plt.subplots(3)
+        fig.set_size_inches(8, 9)
+        legend = []
+        for i in range(0, len(coupling_coefficients["forward_sparam"].columns)):
+            axs[0].plot(
+                coupling_coefficients["position"],
+                coupling_coefficients["forward_sparam"].iloc[:, i],
+                marker=marker_list[i],
+            )
+            legend.append(coupling_coefficients["forward_sparam"].columns[i])
+        axs[0].grid("on")
+        axs[0].set_title("Forward Propagating S-Parameters")
+        axs[0].set_ylabel("Magnitude (dB)")
+        axs[0].set_xlabel("Position (m)")
+
+        legend = []
+        for i in range(0, len(coupling_coefficients["backward_sparam"].columns)):
+            axs[1].plot(
+                coupling_coefficients["position"],
+                coupling_coefficients["backward_sparam"].iloc[:, i],
+                marker=marker_list[i],
+            )
+            legend.append(coupling_coefficients["backward_sparam"].columns[i])
+        axs[1].grid("on")
+        axs[1].set_title("Backward Propagating S-Parameters")
+        axs[1].set_ylabel("Magnitude (dB)")
+        axs[1].set_xlabel("Position (m)")
+
+        legend = []
+        for i in range(0, len(coupling_coefficients["overlap"].columns)):
+            axs[2].plot(
+                coupling_coefficients["position"],
+                coupling_coefficients["overlap"].iloc[:, i],
+                marker=marker_list[i],
+            )
+            legend.append(coupling_coefficients["overlap"].columns[i])
+        axs[2].grid("on")
+        axs[2].set_title("Overlap Coefficients")
+        axs[2].set_ylabel("Magnitude (dB)")
+        axs[2].set_xlabel("Position (m)")
+        fig.legend(legend, loc="center right", bbox_to_anchor=(1.15, 0.5))
+        fig.suptitle(
+            f"Coupling From Mode {input_mode} to Modes Up To Mode {max_coupled_mode}"
+        )
+        fig.tight_layout()
+
+        fig.savefig(
+            str(self.dirpath / f"{self.component.name}_mode_coupling.png"),
+            bbox_inches="tight",
+        )
+
+    def get_neff_vs_position(self, group: int = 1) -> pd.DataFrame:
+        """
+        Get effective index vs position along device
```

Comment:
the device is assumed to be symmetrical across the 2 non-position axes, right (this is just for my understanding)